### PR TITLE
New (real data MRI) examples: scan-specific SSDU, low-field MRI denoising

### PR DESCRIPTION
**Pull request by Andrewwango**
Contribute 2 new examples for self-sup learning for more real MRI use cases.

Fix RAM padding.

### Checks to be done before submitting your PR

- [x] `python3 -m pytest deepinv/tests` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

File: deepinv/models/drunet.py [hunk 1]
```diff
@@ -170,7 +170,10 @@ def __init__(
                     name = "drunet_deepinv_gray_finetune_26k.pth"
                 url = get_weights_url(model_name="drunet", file_name=name)
                 ckpt_drunet = torch.hub.load_state_dict_from_url(
-                    url, map_location=lambda storage, loc: storage, file_name=name
+                    url,
+                    map_location=lambda storage, loc: storage,
+                    file_name=name,
+                    progress=False,
```
Comment (tachella) [inline deepinv/models/drunet.py]: I don't think this is needed now that doctests are removed

File: deepinv/models/ram.py [hunk 2]
```diff
@@ -334,13 +358,7 @@ def forward(
         if physics is None:
             physics = dinv.physics.Denoising(noise_model=dinv.physics.ZeroNoise())
 
-        if img_size is None:
```
Comment (tachella) [inline deepinv/models/ram.py:341]: why removing this? it is supposed to save one A_adjoint call if the user passes img_size

File: deepinv/models/ram.py [hunk 3]
```diff
@@ -334,13 +358,7 @@ def forward(
...
-            if hasattr(physics, "img_shape") and physics.img_shape is not None:
-                img_size = physics.img_shape
-            elif hasattr(physics, "img_size") and physics.img_size is not None:
-                img_size = physics.img_size
-            else:
-                img_size = physics.A_adjoint(y).shape[1:]
+        pad = self.get_pad(physics.A_adjoint(y))
```
Comment (tachella) [inline deepinv/models/ram.py]:
```suggestion
        pad = self.get_pad(img_size)
```

File: deepinv/tests/test_models.py [hunk 4]
```diff
@@ -968,6 +968,18 @@ def __len__(self):
     assert psnr(x_init, x) < psnr(x_hat, x)
 
 
+def test_ram_performance(device):
+    model = dinv.models.RAM(device=device)
```
Comment (tachella) [inline deepinv/tests/test_models.py]: should we test with a less standard image shape? It RAM not already tested elsewhere?

File: deepinv/models/ram.py [hunk 5]
```diff
@@ -284,6 +284,30 @@ def forward_unet(
 
         return x
 
+    def get_pad(self, x: Tensor) -> tuple[int, int, int]:
```
Comment (tachella) [inline deepinv/models/ram.py]: get img_size instead of a tensor

File: examples/self-supervised-learning/demo_lowfieldmri.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+We demonstrate self-supervised (blind) denoising of a low-field MRI scan without ground truth data.
+
+In low-field MRI, images have high noise due to the fixed permanent magnet, and acquiring clean reference images is often impossible.
+One could average the images over multiple repetitions, but if there's any patient motion, the image would be blurry.
+
+Here, we fine-tune the Reconstruct Anything Model (:class:`deepinv.models.RAM`) :footcite:p:`terris2025reconstruct` on a single noisy scan using the self-supervised
+:class:`Recorrupted2Recorrupted loss <deepinv.loss.R2RLoss>` :footcite:p:`pang2021recorrupted`.
+
+Play around with different self-supervised denoising losses (see :ref:`self-supervised-losses`)!
```
Comment (tachella) [inline examples/self-supervised-learning/demo_lowfieldmri.py]:
```suggestion
Play around with different self-supervised denoising losses (see :ref:`self-supervised-losses`) and models!
```

File: examples/self-supervised-learning/demo_lowfieldmri.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+============================================
+
+We demonstrate self-supervised (blind) denoising of a low-field MRI scan without ground truth data.
+
+In low-field MRI, images have high noise due to the fixed permanent magnet, and acquiring clean reference images is often impossible.
+One could average the images over multiple repetitions, but if there's any patient motion, the image would be blurry.
+
+Here, we fine-tune the Reconstruct Anything Model (:class:`deepinv.models.RAM`) :footcite:p:`terris2025reconstruct` on a single noisy scan using the self-supervised
+:class:`Recorrupted2Recorrupted loss <deepinv.loss.R2RLoss>` :footcite:p:`pang2021recorrupted`.
+
```
Comment (tachella) [inline examples/self-supervised-learning/demo_lowfieldmri.py]: maybe I'd add a note saying that better results can be obtained if the model is finetuned on more than a single image?

File: examples/self-supervised-learning/demo_scan_specific.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
+r"""
+Scan-specific zero-shot SSDU for MRI
+=====================================================
+
+We demonstrate scan-specific self-supervised learning, that is, learning to
+reconstruct MRI scans from a single accelerated sample without ground truth.
+
+Here, we demonstrate fine-tuning the Reconstruct Anything Model (:class:`deepinv.models.RAM`) :footcite:p:`terris2025reconstruct`
```
Comment (tachella) [inline examples/self-supervised-learning/demo_scan_specific.py]:
```suggestion
Here, we demonstrate fine-tuning a pretrained model (:class:`deepinv.models.RAM`) :footcite:p:`terris2025reconstruct`
```

File: examples/self-supervised-learning/demo_scan_specific.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+physics = dinv.physics.MultiCoilMRI(
+    device=device, noise_model=dinv.physics.GaussianNoise(0.0)
+)
+
+# %%
+# Model
+# -----
+# For the model we use a fairly large MoDL with a UNet backbone, with 12 iterations:
+
+# denoiser = dinv.models.UNet(2, 2, scales=5, batch_norm=False)
```
Comment (tachella) [inline examples/self-supervised-learning/demo_scan_specific.py]: remove

File: examples/self-supervised-learning/demo_scan_specific.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+    (256, 256), acceleration=6, center_fraction=0.04, rng=rng, device=device
+)
+loss = dinv.loss.mri.WeightedSplittingLoss(
+    mask_generator=mask_generator, physics_generator=physics_generator
+)
+
+# %%
+# Training
+# --------
+# We train the model using the self-supervised loss. We randomly split the dataset into training and validation for
```
Comment (tachella) [inline examples/self-supervised-learning/demo_scan_specific.py]:
```suggestion
# We train the model using the self-supervised loss. We randomly split the volume into training slices and validation slices for
```

File: deepinv/models/ram.py [hunk 11]
```diff
@@ -284,6 +284,30 @@ def forward_unet(
...
+            0,
+            -img_shape[-2] % spatial_pad,
+            -img_shape[-1] % spatial_pad,
+        )
+
+        min_size = 64
+        if img_shape[-2] + pad[1] < min_size:
+            pad = (pad[0], min_size - img_shape[-2], pad[2])
+        if img_shape[-1] + pad[2] < min_size:
+            pad = (pad[0], pad[1], min_size - img_shape[-1])
```
Comment (tachella) [inline deepinv/models/ram.py]: is this for 3d?

File: deepinv/models/ram.py (hunk 2 above)
Comment (Andrewwango) [reply deepinv/models/ram.py:341]: Sorry, reverted

File: deepinv/models/ram.py (hunk 11 above)
Comment (Andrewwango) [reply deepinv/models/ram.py]: No, it's for the channel dim, it's a fix proposed by @matthieutrs

File: deepinv/tests/test_models.py (hunk 4 above)
Comment (Andrewwango) [reply deepinv/tests/test_models.py]: good point, I merged this into the original RAM test. I add this as it's on an actual image, and it's a test that used to be in the RAM doctest (but now longer exists due to the docstring print bug thing) so I'm putting it here instead

File: deepinv/models/ram.py (hunk 11 above)
Comment (Andrewwango) [reply deepinv/models/ram.py]: NOTE: this fix doesn't actually fix a problem of padding. The problem is in dinv MRI img_size handling, now solved.

**Comment by Andrewwango** [comment]
@romainvo if possible I would appreciate a tiny review on just the change to `mri.py`!

File: deepinv/physics/wrappers.py [hunk 12]
```diff
@@ -235,10 +235,20 @@ def A_adjoint(self, y, **kwargs):
         return y
 
     def remove_pad(self, x):
-        return x[..., self.crop[0] :, self.crop[1] :]
+        if len(self.crop) == 2:
+            return x[..., self.crop[0] :, self.crop[1] :]
+        elif len(self.crop) == 3:
+            return x[..., self.crop[0] :, self.crop[1] :, self.crop[2] :]
```
Comment (romainvo) [inline deepinv/physics/wrappers.py:243]: The update to `remove_pad` suggests the operator now supports 3D, in addition to 2D, maybe specify this in the docstring.

File: deepinv/physics/mri.py [hunk 13]
```diff
@@ -359,6 +360,8 @@ def update_parameters(
             mask = (
                 self.check_mask(mask=mask, three_d=self.three_d) if check_mask else mask
             )
+            self.register_buffer("mask", mask)
```
Comment (romainvo) [inline deepinv/physics/mri.py]: The buffer's update is already taken care of in the call to `super().update_parameters` below.

File: deepinv/physics/mri.py [hunk 14]
```diff
@@ -359,6 +360,8 @@ def update_parameters(
...
+            self.img_size = self.mask.shape[1:]
```
Comment (romainvo) [inline deepinv/physics/mri.py]: Should we check if the `self.img_size` changes and if `coil_maps` is also updated in the process ? (like we do in `Downsampling` if the factor is updated but no the filter)

Similary, should we also trigger an `img_size` update if `coil_maps` changes ?

File: deepinv/tests/test_datasets.py [hunk 15]
```diff
@@ -1146,12 +1146,14 @@ def test_FastMRISliceDataset(download_fastmri):
         coil_maps=torch.ones(kspace_shape, dtype=torch.complex64),
         img_size=img_size,
     )
-    rss1 = physics.A_adjoint(kspace1.unsqueeze(0), rss=True, crop=True)
+    rss1 = physics.A_adjoint(kspace1.unsqueeze(0), rss=True)
+    rss1 = physics.crop(rss1, shape=target1.shape)
```
Comment (romainvo) [inline deepinv/tests/test_datasets.py:1150]: Maybe naive question but why do the `physics.crop` outside `A_adjoint`, seems less effective in terms of coverage ? (same in line 1156 and in demo_mri_tour.py)

File: deepinv/physics/wrappers.py [hunk 16]
```diff
@@ -235,10 +235,20 @@ def A_adjoint(self, y, **kwargs):
...
+        else:
+            raise ValueError("Crop must be a tuple of length 2 or 3.")
```
Comment (romainvo) [inline deepinv/physics/wrappers.py]: Would be cleaner to raise the error in the `__init__` I think

File: deepinv/physics/wrappers.py (hunk 12 above)
Comment (Andrewwango) [reply deepinv/physics/wrappers.py:243]: done

File: deepinv/physics/mri.py (hunk 13 above)
Comment (Andrewwango) [reply deepinv/physics/mri.py]: done

File: deepinv/tests/test_datasets.py (hunk 15 above)
Comment (Andrewwango) [reply deepinv/tests/test_datasets.py:1150]: Yeah it's kind of annoying, it's because of the dual purpose of img_size, since it should both be the image size of A^T(y) but also in the fastMRI context the cropped image size. Before, img_size meant the cropped size but this is confusing against deepinv convention. I think the this more explicit way is clearer.

File: deepinv/physics/wrappers.py (hunk 16 above)
Comment (Andrewwango) [reply deepinv/physics/wrappers.py]: done

File: deepinv/physics/mri.py (hunk 14 above)
Comment (Andrewwango) [reply deepinv/physics/mri.py]: Think triggering on coil maps adds too much complexity, but it's a good point and I add a warning if any of them change

File: examples/self-supervised-learning/demo_lowfieldmri.py [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+    ],
+)
+
+# %%
+# Physics
+# -------
+# We define a simple denoising physics with Gaussian noise matching the estimated noise level.
+
+sigma = noise_estimator(y).detach().cpu()
+physics = dinv.physics.Denoising(dinv.physics.GaussianNoise(sigma=sigma))
```
Comment (romainvo) [inline examples/self-supervised-learning/demo_lowfieldmri.py]: GPU docs is failing because physics' default device is "cpu", pass `device=device` to push sigma buffer on device="cuda" (also probably faster noise estimation if `y` is on the gpu).

```suggestion
with torch.no_grad():
    sigma = noise_estimator(y.to(device))
physics = dinv.physics.Denoising(dinv.physics.GaussianNoise(sigma=sigma), device=device)
```

Comment (Andrewwango) [reply examples/self-supervised-learning/demo_lowfieldmri.py]: thanks so much @romainvo !
